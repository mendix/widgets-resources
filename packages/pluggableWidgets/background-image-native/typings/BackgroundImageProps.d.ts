/**
 * This file was generated from BackgroundImage.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, ReactNode } from "react";
import { DynamicValue, NativeImage } from "mendix";

export type ResizeModeEnum = "cover" | "contain" | "stretch" | "center";

export interface BackgroundImageProps<Style> {
    name: string;
    style: Style[];
    image: DynamicValue<NativeImage>;
    resizeMode: ResizeModeEnum;
    opacity: BigJs.Big;
    content?: ReactNode;
}

export interface BackgroundImagePreviewProps {
    class: string;
    style: string;
    image: string;
    resizeMode: ResizeModeEnum;
    opacity: number | null;
    content: { widgetCount: number; renderer: ComponentType };
}
