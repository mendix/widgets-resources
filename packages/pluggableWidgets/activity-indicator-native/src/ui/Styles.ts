import { Style } from "@native-mobile-resources/util-widgets";
import { ViewStyle } from "react-native";

export interface ActivityIndicatorStyle extends Style {
    container: ViewStyle;
    indicator: {
        size: "large" | "small";
        color: string;
    };
}

export const defaultActivityStyle: ActivityIndicatorStyle = {
    container: {},
    indicator: {
        size: "large",
        color: "gray"
    }
};
