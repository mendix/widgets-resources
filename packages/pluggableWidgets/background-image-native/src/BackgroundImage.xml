<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.backgroundimage.BackgroundImage" supportedPlatform="Native" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Background image</name>
    <description />
    <helpUrl>https://github.com/mendix/widgets-resources/blob/master/packages/pluggableWidgets/background-image-native/README.md</helpUrl>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAAHdbkFIAAAABGdBTUEAALGPC/xhBQAAADhlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAAAQKADAAQAAAABAAAAQAAAAABlmWCKAAAHdUlEQVR4Ae0bW2wUVfTso6xgrUjTYJHUUoyo1BeiRSXGVGwwkQ+RthobftQiMfHDR4iJGkI0hkT4kJgKiPGjDWlrDMYPEwONGqJUefigRVIEUh5VDE2BSrPuy3Nm9s7emblz587sLN2Uvcn2Ps77zH2eewugmpqbmzPst28kqZWJNsozeO3DnXxVK4f5lvX9Cb4qQGgo0xp5RBMHRr4+i0h1IQJDpNyk5KZXnuVhHsvMdvLF7iPnMow8xAqEwMqUL6mOajCTkrx5DNmMgOZZkQwRy3dNZJpqIvBQtU7DRBgITjoYfnDygcGBKZXJZEItLS1pVrfk4729vdfxbYYE1kjEom6RhZdjnkEmhmCTFxgTyj8/loSL//Et4rJNA4a26hZHEEPRckcs/nuw7kdtJgcgC2cGXJ9lIonRpi5W03NHH5jRirlmfE+mJI6XfVhuYHU+D4VC7T09Pdv5NpMPiHjeHfc0iDoStWEv3cYTU9n6FRpWvrROwxExsRJT3aSBCMGtzZHBBwfts6CImZABEY+r0dt8oAl5fZE+xYokWtusTrTC4b39CUikAKahrjEb1P4VNJQNPyUgjZNkGHsJ5TQGth5OwriAgc0HNOJmoF5EFIvoOdGtqRcra2ttnBuBR27S+a67z90XNg0YsUBbYZONgRBL0mgygQYLTt+2/i6hnwIg23zC2+SySPGosnIfrkOPOSGYvoEVia2Q/NSy9Z2XYXxs1Ipq1Hnc7HLcaAAFBakCAnxYs+EjUbPvtry7sW/JWUJPHqDtwuHzuZ0YW7DzUcKTB5pqcvpOzxXzkS+ez504VkzLTY5WnPdx2o7jtL0Ud5DLcCepmjzbsflQQtu1rb0rCrOnh+DdnxOQzO4mafLdeCABe0f0Bus+SqSUkgIkgARRoonjYbSy47ekVqdvyPcFVv7zQhp2aRjyP0oKkPC6ijCsvj3n2sdd3Dz/erXupaQAs0puiz+ompr+eCtRuSmgbSNkU6+SFAmSdDEiOtyv7sGsUcJDCorFYnM6OztHpEgl4GR6wLUP4J7gRdyYbEYl6ZDrJfXjYehB3OflVi8BtVSBfDsgyUMlwjIl3CYirffTJqR85ixNf6egAzOO7Yi++HgjnBj8BdCDPyJsCYNbc7d5QMNnwq3E1jqPx06KiCM8rzJaNw8wPCNnFhoNeRaUPJCnDCl5SQFPHqD4FZ3faXMaVPLUCWk7RknfGSdBNSYmU9aTB3hGJy9KJzgeVVr2pAC/FW+vdz/9SyVngZ4+gUrIQUUoj+NJAZ6QLw9fysCng0ltx/z2A2VadIaHy8qePoGI0XHcfpNwStQrKER06B+nsLOGZvrjyQN/T2S088C1SPUGHkKOjqVh51E8DmGiuPrp8QwMjqbhy+Mp+PaM3m6SJqgoK3DgXBq+OqEz/RcNZkcx4vlkbQQWz9adeX4iDFvw0HIhDrYIs0C+WriydyhpCF9+s344oXMgpafqooZwqlficY3OEeWKg0SpDwyM6mO+fWEUltwY1gREkfKZWyNwd5V4T6MasFT+BG8u1kOHZCWlt+5XNFFHd/yrpMBVfTRzdF1QAKVOGJQwEZ+pqwB3ohYF6g1nKI0Ct7OAwU1QwEPJq4Jmo0n6CehobWD6K/RZ76r8sSlRlTwwdT0gXsoV7W1ra6uOx+OdiN6oSBI0Wh9OFG35xOB8OyAbOCqKCz66bPQ72ymtA6LPlo2aGSA+hmQ0YiGfNYT4iKIxtMjRzR1LWV22s7qXXLoOuTAyhQz52BRPR3EqJxiPZy0TDRfjMoEF/Ey6mJBdKr57gAtfA4xvQgK/5jSYB1DIpwcEIH7yWZQcMPnfYHI1KNgcQLHUjt8TMGEJpVKAb+2dZUDX38WQCjYEth22G08Gk0MIViypYA6orXDeY8lgV9oxBRsCFL1vqgH4ZjgJLJhOhtPTk2Lp/uTsgjmAmJOhXq8x/rqcge/OpGECR8mjc0NQi281CpkK6gAvip+8lIauP1KQ4CLrnx0hDimYUx6Cp+dHoPIa52HlRRaPW3AH0IvYvWdT2tuipXPCMDNmNmIIQ/zdQzqcV4xWC7aCnMWw/5Zf9eVk4awQrMCAeFApOE4WjejdUg8axqL4BN6PVwxk2OrbojAWz0DvsZT2zpYnfQKvGhqyVw3UfhAvW74+mesZFKgfGLU/bud5eCkH7oDdwynj5ZhIEfqq9KiYT9QnVtRFYFGVfbxTG/1SeEGx51QKfsi+SuPp8ykH5oAdA0k4hV2VT2V4lbMKx+6CG3TD+k6n4Huc4Fiil9Yr50ehvtI8LBiczyOIQrdg9LuMw6qji4f6LwfmAN74Krykem5BFMe7WTF65E0/el4eQoPczTbTs9qMYG6GNHaBOYC43YtddcW8iOstLX35YkmBOaCQd2iFdJZ91imktCLkXXJAEX6UK6pSqQfk4e5+npaeSXJ3sjwo0DLJIFmWZNLFApNWfS9I2X/z2Y3cG6USCg+kK+BlssexMhV8O4AxJUe0tra+gPnz2LYQf75j9IynS043/gNo8I7u7u5P/BruIqMEvmo88D82wx1NJv+rXQAAAABJRU5ErkJggg==</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="image" type="image">
                    <caption>Image</caption>
                    <description />
                </property>
                <property key="resizeMode" type="enumeration" defaultValue="cover">
                    <caption>Resize mode</caption>
                    <description>Cover: scale to  cover space. Contain: scale to fit in space. Stretch: stretched to fit space. Center: centered and fit in space. For more information see the help page.</description>
                    <enumerationValues>
                        <enumerationValue key="cover">Cover</enumerationValue>
                        <enumerationValue key="contain">Contain</enumerationValue>
                        <enumerationValue key="stretch">Stretch</enumerationValue>
                        <enumerationValue key="center">Center</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="opacity" type="decimal" defaultValue="1.0">
                    <caption>Opacity</caption>
                    <description>Takes a value from 0.0 - 1.0. The lower the value, the more transparent.</description>
                </property>
                <property key="content" type="widgets" required="false">
                    <caption>Content</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name" />
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
