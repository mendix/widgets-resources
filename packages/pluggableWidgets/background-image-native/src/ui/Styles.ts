import { ViewStyle, ImageStyle } from "react-native";
import { Style } from "@native-mobile-resources/util-widgets";

interface CustomImageStyle extends ImageStyle {
    svgColor?: string;
}

export interface BackgroundImageStyle extends Style {
    container: ViewStyle;
    image: CustomImageStyle;
}

export const defaultBackgroundImageStyle: BackgroundImageStyle = {
    container: {},
    image: {}
};
