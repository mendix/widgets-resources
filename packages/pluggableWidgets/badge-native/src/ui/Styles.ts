import { Style } from "@native-mobile-resources/util-widgets";
import { TextStyle, ViewStyle } from "react-native";

export interface BadgeStyle extends Style {
    container: ViewStyle;
    caption: TextStyle;
}

export const defaultBadgeStyle: BadgeStyle = {
    container: {
        alignSelf: "baseline",
        borderRadius: 30,
        paddingHorizontal: 10,
        paddingVertical: 5,
        backgroundColor: "#ccc",
        justifyContent: "center"
    },
    caption: {
        fontSize: 14,
        fontWeight: "bold",
        color: "#444"
    }
};
