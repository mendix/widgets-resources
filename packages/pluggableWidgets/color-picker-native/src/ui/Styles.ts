import { Style } from "@native-mobile-resources/util-widgets";
import { ViewStyle } from "react-native";

export interface ColorPickerStyle extends Style {
    container: ViewStyle;
    preview: ViewStyle;
}

export const defaultColorPickerStyle: ColorPickerStyle = {
    container: {},
    preview: {
        borderRadius: 5,
        minHeight: 50
    }
};
