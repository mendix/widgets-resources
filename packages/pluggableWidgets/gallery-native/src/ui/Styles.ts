import { ViewStyle } from "react-native";

export interface GalleryStyle {
    container?: ViewStyle;
    listStyle?: ViewStyle;
    listItem?: ViewStyle;
    pagination?: ViewStyle;
}

export const defaultGalleryStyle: GalleryStyle = {
    container: {},
    listStyle: {},
    listItem: { flex: 1 },
    pagination: { flexDirection: "row", justifyContent: "center" }
};
