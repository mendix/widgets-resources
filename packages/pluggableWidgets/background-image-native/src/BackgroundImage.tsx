import { createElement } from "react";
import { View, StyleSheet } from "react-native";
import { ValueStatus } from "mendix";
import { Image } from "mendix/components/native/Image";
import { flattenStyles } from "@native-mobile-resources/util-widgets";

import { BackgroundImageStyle, defaultBackgroundImageStyle } from "./ui/Styles";
import { BackgroundImageProps } from "../typings/BackgroundImageProps";

export function BackgroundImage(props: BackgroundImageProps<BackgroundImageStyle>): JSX.Element | null {
    const styles = flattenStyles(defaultBackgroundImageStyle, props.style);
    const { image, name, resizeMode } = props;
    let opacity = Number(props.opacity.toFixed());

    if (opacity < 0) {
        opacity = 0;
    } else if (opacity > 1) {
        opacity = 1;
    }

    if (image.status !== ValueStatus.Available || !image.value) {
        return null;
    }

    const imageStyle = [
        StyleSheet.absoluteFill,
        typeof image.value === "number"
            ? { width: undefined, height: undefined }
            : typeof image.value === "string"
            ? { width: "100%", height: "100%" }
            : undefined,
        styles.image,
        { opacity, resizeMode }
    ];

    return (
        <View style={styles.container} testID={name}>
            <Image source={image.value} style={imageStyle} color={styles.image.svgColor} testID={`${name}$image`} />
            {props.content}
        </View>
    );
}
