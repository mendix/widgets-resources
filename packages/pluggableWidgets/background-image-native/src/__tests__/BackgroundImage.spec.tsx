import { createElement } from "react";
import { Text } from "react-native";
import { render } from "react-native-testing-library";
import { dynamicValue } from "@native-mobile-resources/util-widgets";

import { BackgroundImage } from "../BackgroundImage";
import { BackgroundImageProps } from "../../typings/BackgroundImageProps";
import { BackgroundImageStyle } from "../ui/Styles";
import { NativeImage } from "mendix";
import Big from "big.js";

jest.mock("mendix/components/native/Image", () => jest.requireActual("./__mocks__/mendix/components/native/Image"));

const defaultProps: BackgroundImageProps<BackgroundImageStyle> = {
    name: "backgroundImageTest",
    style: [],
    image: dynamicValue<NativeImage>({ uri: "path/to/image" }),
    resizeMode: "cover",
    opacity: new Big(0.3333),
    content: <Text>Content</Text>
};

describe("BackgroundImage", () => {
    it("renders with default styles", () => {
        const component = render(<BackgroundImage {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders with custom styles", () => {
        const style: BackgroundImageStyle[] = [
            {
                container: { height: "50%" },
                image: { width: "100%", height: "100%" }
            },
            {
                container: { height: "80%" },
                image: {}
            }
        ];

        const component = render(<BackgroundImage {...defaultProps} style={style} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders nothing when image is loading for the first time", () => {
        const image = dynamicValue<NativeImage>(undefined, true);

        const component = render(<BackgroundImage {...defaultProps} image={image} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders previous image when image is reloading", () => {
        const image = dynamicValue<NativeImage>({ uri: "path/to/image" }, true);

        const component = render(<BackgroundImage {...defaultProps} image={image} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders content only when image is unavailable", () => {
        const image = dynamicValue<NativeImage>({ uri: "path/to/image" });

        const component = render(<BackgroundImage {...defaultProps} image={image} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders without content", () => {
        const content = null;

        const component = render(<BackgroundImage {...defaultProps} content={content} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("use correct opacity when image opacity is out of range", () => {
        const negativeImageOpacity = new Big(-0.333);
        const positiveImageOpacity = new Big(1.333);

        const component1 = render(<BackgroundImage {...defaultProps} opacity={negativeImageOpacity} />);
        expect(component1.toJSON()).toMatchSnapshot();
        const component2 = render(<BackgroundImage {...defaultProps} opacity={positiveImageOpacity} />);
        expect(component2.toJSON()).toMatchSnapshot();
    });
});
