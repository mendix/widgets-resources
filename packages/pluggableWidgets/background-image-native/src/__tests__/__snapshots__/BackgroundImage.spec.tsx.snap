// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage renders content only when image is unavailable 1`] = `
<View
  style={Object {}}
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
        Object {
          "opacity": 0.3333,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
  <Text>
    Content
  </Text>
</View>
`;

exports[`BackgroundImage renders nothing when image is loading for the first time 1`] = `null`;

exports[`BackgroundImage renders previous image when image is reloading 1`] = `null`;

exports[`BackgroundImage renders with custom styles 1`] = `
<View
  style={
    Object {
      "height": "80%",
    }
  }
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {
          "height": "100%",
          "width": "100%",
        },
        Object {
          "opacity": 0.3333,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
  <Text>
    Content
  </Text>
</View>
`;

exports[`BackgroundImage renders with default styles 1`] = `
<View
  style={Object {}}
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
        Object {
          "opacity": 0.3333,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
  <Text>
    Content
  </Text>
</View>
`;

exports[`BackgroundImage renders without content 1`] = `
<View
  style={Object {}}
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
        Object {
          "opacity": 0.3333,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
</View>
`;

exports[`BackgroundImage use correct opacity when image opacity is out of range 1`] = `
<View
  style={Object {}}
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
        Object {
          "opacity": 0,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
  <Text>
    Content
  </Text>
</View>
`;

exports[`BackgroundImage use correct opacity when image opacity is out of range 2`] = `
<View
  style={Object {}}
  testID="backgroundImageTest"
>
  <Image
    source={
      Object {
        "uri": "path/to/image",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
        Object {
          "opacity": 1,
          "resizeMode": "cover",
        },
      ]
    }
    testID="backgroundImageTest$image"
  />
  <Text>
    Content
  </Text>
</View>
`;
